verifier/src/report_body.rs

```diff
@@ -2,72 +2,95 @@
 use subtle::CtOption;
 
 /// Verify the report body is as expected.
+#[derive(Clone, Debug, Eq, Hash, PartialEq, Default)]
+pub struct ReportBodyVerifier {
+    attributes_verifier: ReportBodyVerifierKind,
+    config_id_verifier: ReportBodyVerifierKind,
+}
+
 #[derive(Clone, Debug, Eq, Hash, PartialEq)]
-pub struct ReportBodyVerifier<'a> {
-    expected_report_body: ReportBody,
-    report_body: &'a ReportBody,
+pub enum ReportBodyVerifierKind {
+    AttributesVerifier(AttributesVerifier),
+    ConfigIdVerifier(ConfigIdVerifier),
+    AlwaysTrue(AlwaysTrue),
 }
 
-impl<'a> ReportBodyVerifier<'a> {
-    /// Create a new [`ReportBodyVerifier`] instance.
-    ///
-    /// # Arguments:
-    /// * `expected_report_body` - The expected report body.
-    /// * `report_body` - The report body to verify conforms to the
-    ///   `expected_report_body`.
-    pub fn new(expected_report_body: ReportBody, report_body: &'a ReportBody) -> Self {
-        Self {
-            expected_report_body,
-            report_body,
+impl Verifier<ReportBody> for ReportBodyVerifierKind {
+    type Error = VerificationError;
+
+    fn verify(&self, evidence: &ReportBody) -> CtOption<Self::Error> {
+        match self {
+            Self::AttributesVerifier(verifier) => verifier.verify(evidence),
+            Self::ConfigIdVerifier(verifier) => verifier.verify(evidence),
+            Self::AlwaysTrue(verifier) => verifier.verify(evidence),
         }
     }
 }
 
-impl Verifier for ReportBodyVerifier<'_> {
+impl Default for ReportBodyVerifierKind {
+    fn default() -> Self {
+        Self::AlwaysTrue(AlwaysTrue::default())
+    }
+}
+
+impl ReportBodyVerifier {
+    /// Verify the report body against the provided Attributes
+    pub fn with_attributes(mut self, attributes: Attributes) -> Self {
+        self.attributes_verifier =
+            ReportBodyVerifierKind::AttributesVerifier(AttributesVerifier::new(attributes));
+        self
+    }
+
+    /// Verify the report body against the provided ConfigId
+    pub fn with_config_id(mut self, config_id: ConfigId) -> Self {
+        self.config_id_verifier =
+            ReportBodyVerifierKind::ConfigIdVerifier(ConfigIdVerifier::new(config_id));
+        self
+    }
+}
+
+impl Verifier<ReportBody> for ReportBodyVerifier {
     type Error = AndError<VerificationError, VerificationError>;
 
-    fn verify(&self) -> CtOption<Self::Error> {
-        let attributes_verifier =
-            AttributesVerifier::new(self.expected_report_body.attributes(), self.report_body);
-        let config_id_verifier =
-            ConfigIdVerifier::new(self.expected_report_body.config_id(), self.report_body);
-        let and_verifier = And::new(attributes_verifier, config_id_verifier);
-        and_verifier.verify()
+    fn verify(&self, evidence: &ReportBody) -> CtOption<Self::Error> {
+        let verifier = self
+            .attributes_verifier
+            .clone()
+            .and(self.config_id_verifier.clone());
```

Comment:
Not fond of this cloning, it may be that I need to use refernces and lifetimes here.

Reply:
Could `.and()` be refactored to take `&self` rather than `self`? I realize that would also change the signature of `And<R, L>` so it might be outside of the scope of this PR.

Reply:
Possibly, I'm a little reluctant as the idea is to build up a tree of verifiers and if they're references then one will need to keep all of the tree nodes around instead of passing the tree around and forgetting about it.